Introduce overloads for getting a goto model out of load_java_class

thk123 opened the pull request:
This can be used when you need both the symbol table and the goto functions map, or when you have whole goto_model passes.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [n/a] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Introduce overloads for getting a goto model out of load_java_class

This can be used when you need both the symbol table and the goto
functions map, or when you have whole goto_model passes.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ae0a629db7a811464a6c60ee99ded5bdb88dcd02).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86029107

romainbrenguier commented on jbmc/unit/java-testing-utils/load_java_class.cpp:
```diff
@@ -89,6 +79,22 @@ symbol_tablet load_java_class(
+    ->get_symbol_table();
+}
+
+std::unique_ptr<goto_modelt> load_goto_model_from_java_class(
```
This should be documented (maybe by linking to another method) :dog2:

romainbrenguier commented on jbmc/unit/java-testing-utils/load_java_class.cpp:
```diff
@@ -173,3 +179,22 @@ symbol_tablet load_java_class(
     java_class_name, class_path, main, std::move(java_lang), command_line);
 }
+
+std::unique_ptr<goto_modelt> load_goto_model_from_java_class(
```
:dog2:

romainbrenguier commented on jbmc/unit/java-testing-utils/load_java_class.cpp:
```diff
@@ -173,3 +179,22 @@ symbol_tablet load_java_class(
+    java_class_name,
+    class_path,
+    main,
+    get_language_from_mode(ID_java),
```
:question: can't we pass the new language directly instead of registering it and then getting it?

peterschrammel replied:
At the moment you can't, but the whole language registration thing will be removed anyway soonish...

thk123 replied:
I can (in fact elsewhere it is done when providing a custom language):
```C++
std::unique_ptr<languaget> lang = new_java_bytecode_language();
..
main,
std::move(lang),
```

Have done this as reduces dependencies on external code so unless there is a reason this is wrong @peterschrammel ?

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ca0ecece1c6891508ed6c186f1e9f1ed508ea62c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86139315

kroening commented:
Why not simply return a goto_modelt?

thk123 commented:
@kroening As in get rid of all the methods returning `symbol_tablet`? Or return `goto_modelt` instead of a `std::unique_ptr<goto_modelt>`?

The former because I don't want to update all the tests - not  a great reason, but I'm not sure providing useful overloads is such a bad thing anyway. 
The later because I didn't think I could? Can try it out if that's what you meant.

kroening commented:
@thk123 The latter, please, and yes, you can!

Commit:
Structure the documentation in load_java_class

Put the main documention on the method that does the work. Have
overloads reference the methods they call with the specialisation they
provide.

thk123 commented:
@kroening I do not seem to be able to:
```
error: use of deleted function ‘goto_modelt::goto_modelt(const goto_modelt&)’
   return *(maybe_goto_model.get());
```

Commit:
Don't register languages in load_java_class

Is simpler and possible to just create and pass in.

thk123 commented:
@romainbrenguier changes applied

kroening commented:
@thk123 Use an rvalue! This way, the expensive copy is avoided.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: e1979271fc16c9532b9e6f0ce3f5322fe807fffa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86278206

Commit:
Move the goto model directly back

thk123 commented:
@kroening figured it out thanks!

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 4b8a65562e6439281dafb9cb06a5bc21c3dde691).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86293756
